fix #19636: Beam job change user null bio to empty string in UserSettings model

Somnath-Mishra opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #19636.
2. This PR does the following: beam job change user null bio to empty string. Some user_bio fields on the server are null. This results in the entire server throwing a 500 for some users.
3. (For bug-fixing PRs only) The original bug occurred because: when [this PR](https://github.com/oppia/oppia/pull/19399) is deployed.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

- [x] A testing doc has been written: [Testing Doc](https://docs.google.com/document/d/1UeKqoh_BuNbUNM3WWqHCV5ZDB56wYRJgSaPeBoSTadA/edit?usp=sharing)
- [ ] _(To be confirmed by the server admin)_ All jobs in this PR have been verified on a live server, and the PR is safe to merge.

## Proof that changes are correct

[change user null bio to empty string audit and change data job.zip](https://github.com/user-attachments/files/18560732/change.user.null.bio.to.empty.string.audit.and.change.data.job.zip)



<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Somnath-Mishra committed:
fix #19636: Change user_bio null to empty string using beam job

Somnath-Mishra committed:
add test user with null bio, short bio, long bio

Somnath-Mishra committed:
update assert statement message in test file

Somnath-Mishra committed:
remove unwanted packages

Somnath-Mishra committed:
fix linter and mypy type checks

Somnath-Mishra commented:
@lkbhitesh07 @U8NWXD PTAL

Somnath-Mishra committed:
remove unwanted result printing in output section

Somnath-Mishra commented:
@JayVivarekar PTAL

lkbhitesh07 commented on core/jobs/batch_jobs/user_bios_change_jobs_test.py:
```diff
@@ -0,0 +1,122 @@
+        assert user_setting_model is not None
+        self.assertTrue(
+            isinstance(user_setting_model.user_bio, str),
+            """user_bio is not type of string"""
```
Please make sure not to use the wrong help text, it initiates the confusion around the changes. Please correct here and elsewhere.
This was the case in your previous prs as well, please make sure to follow the review comments in your future changes.
Thanks

Somnath-Mishra replied:
https://github.com/oppia/oppia/issues/19636#issuecomment-2602600069
https://github.com/oppia/oppia/issues/19636#issuecomment-2603014977
@lkbhitesh07 , could you please read these comments? I think something is conflicting with the conventions.

lkbhitesh07 replied:
Hmm, okay, I think let's keep the way it is.

Somnath-Mishra replied:
Ok.

lkbhitesh07 reviewed (changes requested):
Thanks @Somnath-Mishra, I have just one inline comment.

Somnath-Mishra commented:
@lkbhitesh07 PTAL

lkbhitesh07 commented on core/jobs/batch_jobs/user_bios_change_jobs_test.py:
```diff
@@ -0,0 +1,122 @@
+            """user_bio is not type of string"""
+        )
+
+    def test_user_with_short_bio(self) -> None:
```
What is the purpose of this and the below test, I am not able to understand? Is there anything I am missing?

Somnath-Mishra replied:
Here suggested https://github.com/oppia/oppia/issues/19636#issuecomment-2585144001

lkbhitesh07 replied:
But I think the job is not doing anything for the bio which is too large or too small, right? It is only fixing the bio having null values. Could you please confirm on the issue thread regarding the action that needs to be performed in case of long and short bio?

Somnath-Mishra replied:
As user_bio with a NULL value or more than 2000 characters ( https://github.com/oppia/oppia/issues/19694 ) is invalid, I initially planned to write a Beam job to handle both cases. However, as Sean Lip suggested focusing on the NULL value part first, he initially recommended writing tests for users with both short and long bios. But since this PR only handles the NULL case, it is sufficient to test users with NULL and non-NULL bios. I have updated the code accordingly.

lkbhitesh07 reviewed (changes requested):
Hi, Just one clarification question, also, I think the proof video you have added does not align with the changes you have added. Let's update the proof section.

**EDIT** Please ignore the last section of this comment.

Somnath-Mishra commented:
> Just FYI, I have added one comment to the testing doc @Somnath-Mishra, also please note that I have updated my comment [here](https://github.com/oppia/oppia/pull/21772#pullrequestreview-2575688453). Thanks

Do I have to write AuditChangeUserNullBiosToEmptyStringJob again as the third job in the testing document?

lkbhitesh07 commented:
Yes, that would help us in verifying that all the null values have been updated.

Somnath-Mishra commented:
> Yes, that would help us in verifying that all the null values have been updated.

Done.

Somnath-Mishra commented:
@lkbhitesh07 PTAL

lkbhitesh07 reviewed (changes requested):
Hi @Somnath-Mishra, I have replied to one of our active conversations. Also, please note that I have added some comments on to your testing doc.
Thanks

Somnath-Mishra committed:
correct test function name

Somnath-Mishra commented:
@lkbhitesh07  I have updated and replied in the testing doc as well as here. PTAL

Somnath-Mishra committed:
fix linter extra new line at the end of file

Somnath-Mishra committed:
Merge branch 'develop' into runAuditJobsOnTheServerForValidationsOfUserSettings

lkbhitesh07 commented:
Hi @Somnath-Mishra, Your last changes looks good, I think Sean have made some comments on your doc. Please assign me again once you update the doc.
Thanks

Somnath-Mishra commented:
@lkbhitesh07 , I have updated the document. Please review it and let me know if any further updates are needed. PTAL.
Thanks

kevintab95 commented:
Hi @Somnath-Mishra,  thanks for the PR! Given that this PR introduces a new Beam job, please follow instructions [here](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api) to test the job against the server if you haven't done so already.

lkbhitesh07 commented:
Thanks @Somnath-Mishra, I have reviewed the doc and have left some review comments.

Somnath-Mishra commented:
@lkbhitesh07 I asked a question in docs.PTAL
Thanks

lkbhitesh07 commented:
There are still some changes require in the doc as requested by Sean, I am going to unassign myself for now.
Thanks

Somnath-Mishra commented:
> There are still some changes require in the doc as requested by Sean, I am going to unassign myself for now. Thanks

I have already updated the document as suggested by Sean. If anything is still missing or needs to be added, please let me know. Sorry for explicitly asking. Thanks for the review

JayVivarekar reviewed (changes requested):
Leaving a blocking review until this is tested.

Somnath-Mishra committed:
Merge branch 'develop' into runAuditJobsOnTheServerForValidationsOfUserSettings

Somnath-Mishra committed:
Merge branch 'develop' into runAuditJobsOnTheServerForValidationsOfUserSettings

lkbhitesh07 commented:
Confirmed with @JayVivarekar that the doc looks good, approving it now.
Thanks

Somnath-Mishra commented:
@JayVivarekar Please run this job at backup and production server. Thanks.

Somnath-Mishra committed:
Merge branch 'develop' into runAuditJobsOnTheServerForValidationsOfUserSettings

Somnath-Mishra committed:
Merge branch 'develop' into runAuditJobsOnTheServerForValidationsOfUserSettings

JayVivarekar commented:
@Somnath-Mishra The backup server is currently in use right now. I will test the PR as soon as the server is free.

JayVivarekar reviewed (approved):
@Somnath-Mishra This PR has been tested successfully, let's merge it!

Somnath-Mishra commented:
> @Somnath-Mishra This PR has been tested successfully, let's merge it!

Please merge it.
Thanks.

Somnath-Mishra commented:
@seanlip Please merge this PR.
Thanks.

Somnath-Mishra commented:
The flake occurred in the merge queue here: [Link](https://github.com/oppia/oppia/actions/runs/13846009031/job/38745325468). However, this test passed in a previous run: [Link](https://github.com/oppia/oppia/actions/runs/13612186310/job/38050930956?pr=21772)

Somnath-Mishra commented:
Failed again due to a flake. I have mentioned it here: https://github.com/oppia/oppia/issues/20640#issuecomment-2723761750

Somnath-Mishra commented:
@seanlip Please force-merge it. Failed again due to a flake. I have mentioned it here: https://github.com/oppia/oppia/issues/20622#issuecomment-2723951307
However, this test passed in a previous run: [link](https://github.com/oppia/oppia/actions/runs/13612186310/job/38050927909?pr=21772)
Wiki [referrence](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR#what-to-do-if-ci-checks-only-fail-in-the-merge-queue)
Thanks.

Somnath-Mishra commented:
https://github.com/oppia/oppia/issues/21292#issuecomment-2723445180 here I mentioned in issue thread. (occurred in https://github.com/oppia/oppia/pull/21772#event-16750934496)

seanlip commented:
OK, thanks @Somnath-Mishra. I confirm that the flakes are valid. Force-merging per procedure at https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR#what-to-do-if-ci-checks-only-fail-in-the-merge-queue -- thanks!
